### ISSUE TITLE
there should be no difference between "binder-unreachable"  and "type-unreachable" code

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#97]
original issue: https://github.com/microsoft/pyright/issues/3639

typescript does the same thing, but i dont think this implementation detail should be visible to the user in any way

> **Describe the bug** if a variable with type `Literal[False]` is checked in an `if` statement, it's incorrectly marked as reachable
> 
> **Expected behavior** code within the `if` statement is marked as unreachable
> 
> **Screenshots or Code** ![image](https://user-images.githubusercontent.com/57028336/176593845-7ea05fb7-0cf6-447d-8c2a-5082ac9f76ce.png)
> 
> ```python
> if False:
>     print(1)  # pyright unreachable
> 
> foo: Literal[False] = False
> if foo:
>     print(1)  # not unreachable
> ```
> 
> **VS Code extension or command-line** vscode

**Comment by KotlinIsland** [comment]
Super expressions of `Never` moments, should be counted as unreachable
```py
print(  # expect error, unreachable
    exit()
)
```

**Comment by phw** [comment]
@DetachHead 

Continuing the discussion from #7 and my suggestion to separate the aspects

There is a big qualitative difference between code that is unreachable due to the code structure and code that should not be reached based on type hints.

See this example:

```python
def my_func():
    if True:
        return do_something()
    return do_something_else()
```

`do_something_else` is literally unreachable. This code **can not** ever be reached when running the program. And it is possible to statically determine this, hence having basedpyright warn on this is, I think, universally useful.

But this is different:

```python
def my_func(v: str | bytes):
    if isinstance(c, bytes):
        return v.decode()
    elif isinstance(c, str):
        return v
    raise TypeError("Unexpected type, expected string or bytes")
```

Based on the type hints, `raise TypeError` **should** never be reached. But it can be reached, e.g. by calling it with `my_func(None)`.

The example is maybe a bit simplistic, but the example would even treat this condition as a runtime error. Presumably it is very important to this program to only have the allowed types used. A static type checker cannot guarantee the code not being reached. Actually reporting "code is unreachable" is arguably a lie.

This is not a type checker implementation detail, it is a qualitative difference. Hence I think it should be separate configuration parameters for each of those conditions.

**Comment by Paillat-dev** [comment]
Personally, I can add that I have been having the latter issue especially when working on libraries, where you can't realistically expect your users to all type-check their code and in some cases would like to still provide some helpful error messages.

Similar issue with reportunnecessaryisinstance, which we simply disabled.

**Comment by phw** [comment]
Yes, actually this. If `my_func` from the example would be a public exposed function that kind of runtime type check could make sense, and the unreachable code warning is wrong. Of course you can litter the code with type ignores, but then you probably don't catch the case where the code actually becomes unreachable after some change.

**Comment by DetachHead** [comment]
fair enough, thanks for the examples. i guess they should be considered two very different concepts when working on a library so it arguably does make sense to report them as different diagnostics.

though i feel very strongly that *everyone* should *always* be type checking their code (personally i don't even entertain the idea that someone can call my api with the wrong types lol). what i would recommend for your case however is to use a runtime type checking library such as pydantic's [`@validate_call` decorator](https://docs.pydantic.dev/latest/concepts/validation_decorator/)

anyway, i'm going to close this issue as it's proposing the exact opposite behavior to what you guys are suggesting. feel free to open a new issue to discuss splitting up the rule into two if you aren't satisfied with the `validate_call` solution

**Comment by phw** [comment]
> feel free to open a new issue to discuss splitting up the rule into two if you aren't satisfied with the validate_call solution

I haven't created the opposite ticket, yet. But to be honest, I have given up on trying to make basedpyright work for me. It seems to be already incompatible with my usage on a philosophical level, and making it work as I need it requires disabling checks to a level that makes it feel like I'm fighting the intended use. Hence I have concluded that me and basedpyright just don't fit 🤷
